### ISSUE TITLE
Crash when spamming Cmd+R

### ISSUE THREAD

**Issue by hartbit** [schibsted/layout#9]
Here's the stack trace:

```
#0	0x000000010e074420 in specialized _assertionFailure(_:_:file:line:flags:) ()
#1	0x000000010ca2753c in LayoutNode.cleanUp() at /Users/david/Downloads/layout-master/Layout/LayoutNode.swift:625
#2	0x000000010ca547a2 in LayoutNode.unbind() at /Users/david/Downloads/layout-master/Layout/LayoutNode.swift:1504
#3	0x000000010ca5477c in LayoutNode.unbind() at /Users/david/Downloads/layout-master/Layout/LayoutNode.swift:1502
#4	0x000000010ca5477c in LayoutNode.unbind() at /Users/david/Downloads/layout-master/Layout/LayoutNode.swift:1502
#5	0x000000010ca5477c in LayoutNode.unbind() at /Users/david/Downloads/layout-master/Layout/LayoutNode.swift:1502
#6	0x000000010ca4fe5c in LayoutNode.unmount() at /Users/david/Downloads/layout-master/Layout/LayoutNode.swift:1412
#7	0x000000010c95e53f in LayoutViewController.layoutNode.didset at /Users/david/Downloads/layout-master/Layout/LayoutViewController.swift:14
#8	0x000000010c95e32c in LayoutViewController.layoutNode.setter ()
#9	0x000000010c61922d in ExamplesViewController.layoutNode.setter ()
#10	0x000000010c965fc0 in protocol witness for LayoutLoading.layoutNode.setter in conformance LayoutViewController ()
#11	0x000000010c92b1df in closure #1 in LayoutLoading.reloadLayout(withCompletion:) at /Users/david/Downloads/layout-master/Layout/LayoutLoading.swift:63
#12	0x000000010c92b6d7 in partial apply for closure #1 in LayoutLoading.reloadLayout(withCompletion:) ()
#13	0x000000010c97f90e in closure #1 in LayoutLoader.loadLayoutNode(withContentsOfURL:relativeTo:state:constants:completion:) at /Users/david/Downloads/layout-master/Layout/LayoutLoader.swift:165
#14	0x000000010c98b5b8 in partial apply for closure #1 in LayoutLoader.loadLayoutNode(withContentsOfURL:relativeTo:state:constants:completion:) ()
#15	0x000000010c97a94e in didComplete #1 () in Layout.processTemplates(completion:) at /Users/david/Downloads/layout-master/Layout/LayoutLoader.swift:72
#16	0x000000010c979b69 in Layout.processTemplates(completion:) at /Users/david/Downloads/layout-master/Layout/LayoutLoader.swift:101
#17	0x000000010c985e10 in processLayoutData #1 (_:) in LayoutLoader.loadLayout(withContentsOfURL:relativeTo:completion:) at /Users/david/Downloads/layout-master/Layout/LayoutLoader.swift:232
#18	0x000000010c9841c3 in LayoutLoader.loadLayout(withContentsOfURL:relativeTo:completion:) at /Users/david/Downloads/layout-master/Layout/LayoutLoader.swift:277
#19	0x000000010c97eb32 in LayoutLoader.loadLayoutNode(withContentsOfURL:relativeTo:state:constants:completion:) at /Users/david/Downloads/layout-master/Layout/LayoutLoader.swift:170
#20	0x000000010c9802b8 in LayoutLoader.reloadLayoutNode(withCompletion:) at /Users/david/Downloads/layout-master/Layout/LayoutLoader.swift:185
#21	0x000000010c92b08b in LayoutLoading.reloadLayout(withCompletion:) at /Users/david/Downloads/layout-master/Layout/LayoutLoading.swift:69
#22	0x000000010c95fede in LayoutViewController._reloadLayout() at /Users/david/Downloads/layout-master/Layout/LayoutViewController.swift:63
```

It triggers `cleanUp`'s `assert(_evaluating.isEmpty)`.

**Comment by nicklockwood** [comment]
Thanks, I've seen this a couple of times, but it's hard to repro reliably. I'm fairly sure it's a re-entrancy issue of some sort.

**Comment by nicklockwood** [comment]
OK, so this was a fun bug.

When Layout begins evaluating an expression it pushes the expression name onto the `_evaluating` stack, then pops it off again when it's done (this is done in order to detect infinite recursion).

The assertion that was failing checks that the `_evaluating` stack is empty when reloading the layout. Evaluation is done synchronously on the main thread, and pushes/pops of the `_evaluating` stack are supposed to be balanced, so it should be impossible to reload the app in between an expression being pushed onto the stack and it being popped off again.

The assertion was throwing because sometimes the `_evaluating` stack contained an `attributedText` expression when the LayoutViewController was reloaded. It turns out that this was due to the HTML parsing feature of Layout's attributedString expressions.

Layout uses NSMutableAttributedString's `init(data:options:documentAttributes:)` method with the option `[NSDocumentTypeDocumentAttribute: NSHTMLTextDocumentType]` to parse HTML. As you might guess, this used WebKit under the hood.

What you might not expect is that when this method initializes WebKit, it spins up a new runloop instance, which causes iOS to process input, including any queued UIKeyCommand events.

So even though expression evaluation and UIKeyCommand both occur on the main thread, it's possible for Layout to receive a reload event *inside* `NSMutableAttributedString.init` while it's evaluating an attributed string expression, and that is how the assertion ends up getting called.

The workaround I've found is to async dispatch the reload event from the keyboard handler, so that the attributed string expression can finish processing and be popped off the `_evaluating` stack before the reload occurs.

<img width="364" alt="screen shot 2017-08-10 at 10 23 48" src="https://user-images.githubusercontent.com/546885/29164345-034207ae-7db7-11e7-9d9e-4bc91a3c28e1.png">
